pages/intro.md

```diff
@@ -8,13 +8,644 @@ hideInToc: true
+let state = { id: 0, todos: [] }
+
+mk('div', { id: 'app' }, [
+  mk('form', null, [
+    (ui.input = mk('input')),
+    (ui.add = mk('button', { onclick: add }, ['Add ToDo'])),
+  ]),
+  (ui.todos = mk('ul', { style: 'padding:5px;' })),
+])
+```
+
+```js
+function app() {
+  let ui = {}
+  let state = { id: 0, todos: [] }
+
+  return mk('div', { id: 'app' }, [
+    mk('form', null, [
+      (ui.input = mk('input')),
+      (ui.add = mk('button', { onclick: add }, ['Add ToDo'])),
+    ]),
+    (ui.todos = mk('ul', { style: 'padding:5px;' })),
+  ])
+}
+```
+
+```js
+function render() {
+  document.body.append(app())
+}
+
+render()
+```
+
+```js
+function app() {
+  ...
+
+  function createTodo(todo) {
+    let item, text, x;
+
+    item = mk('li', { style: 'display:flex;' }, [
+      (text = mk('span', { style: 'flex:1;', ondblclick: edit }, [todo.text])),
+      (x = mk('button', { onclick: remove }, ['X'])),
+    ]);
+  }
+
+  return item;
+}
+```
+
+```js
+// we need to create a edit, remove and add function
+function add(e) {
+  e.preventDefault()
+
+  const text = ui.input.value.trim()
+  if (!text) return
+
+  ui.input.value = ''
+  const todo = { id: state.id++, text, completed: false }
+
+  state.todos.push(todo)
+  ui.todos.append(createTodo(todo))
+}
+```
+
+```js
+function remove() {
+  state.todos = state.todos.filter((todo) => todo.id !== id)
+  item.remove()
+}
+```
+
+```js
+function edit() {
+  let editing = mk('input', {
+    style: 'flex: 1;',
+    value: todo.text,
+    onblur: cancel,
+    onkeydown,
+  })
+
+  text.replaceWith(editing)
+  editing.focus()
+  x.disabled = true
+  x.style = 'pointer-events: none;'
+}
+```
+
+```js
+// we need to create a cancel and onkeydown function
+function cancel() {
+  x.disabled = editing.replaceWith(text)
+}
+
+function onkeydown(e) {
+  switch (e.keycode) {
+    case 13:
+      text.textContent = todo.text = editing.value.trim()
+    case 27:
+      editing.blur()
+  }
+}
+```
+
+```js
+// combined onkeydown, cancel to be in edit function
+function edit() {
+  function onkeydown(e) {...}
+  function cancel() {...}
+
+  let editing = mk('input', { style: 'flex: 1;', value: todo.text, onblur: cancel, onkeydown })
+
+  text.replaceWith(editing)
+  editing.focus()
+  x.disabled = true
+  x.style = 'pointer-events: none;'
+}
+```
+
+```js
+// put remove and edit function into createTodo
+function createTodo(todo) {
+  let item, text, x;
+
+  item = mk('li', { style: 'display:flex;' }, [
+    (text = mk('span', { style: 'flex:1;', ondblclick: edit }, [todo.text])),
+    (x = mk('button', { onclick: () => remove(todo.id) }, ['X'])),
+  ]);
+
+  function remove(id) {...}
+
+  function edit() {
+    function onkeydown(e) {...}
+    function cancel() {...}
+
+    let editing = mk('input', { style: 'flex: 1;', value: todo.text, onblur: cancel, onkeydown })
+    text.replaceWith(editing)
+    editing.focus()
+    x.disabled = true
+    x.style = 'pointer-events: none;'
+  }
+  return item;
+}
```

Comment:
Great unit transitions here for better explanation. But I think, since the title is "From pure JavaScript to React", we need to create the Pure react version of this function and transition to it immediately so that the difference can be compared instantly. I mean the createTodo function is already React-like, but it is not fully react yet. So I suggest you write this exact function in react and then the next click should show that full react CreateTodo component of this function (no need to explain again, just for comparison). It's a suggestion, not an issue.

Reply:
I created the React implementation and created a running app below the magic move.